File: 801-is-graph-bipartite/is-graph-bipartite.cpp
class Solution {
public:
    bool dfs(int curr, vector<int>& color, int currcolor, vector<vector<int>>& graph){
        color[curr]=currcolor;
        for(int nei:graph[curr]){
            if(color[nei]==currcolor){
                return false;
            }
            if(color[nei]==-1){
                int neicolor=1-currcolor;
                if(dfs(nei,color,neicolor,graph)==false){
                    return false;
                }
            }
        }
        return true;
    }
    bool isBipartite(vector<vector<int>>& graph) {
        int v=graph.size();
        vector<int>color(v,-1);
        for(int i=0;i<v;i++){
            if(color[i]==-1){
                if(dfs(i,color,0,graph)==false){
                    return false;
                }
            }
            
        }
        return true;
    }
};

File: 801-is-graph-bipartite/is-graph-bipartite.py
from collections import deque
from typing import List

class Solution:
    def isBipartite(self, graph: List[List[int]]) -> bool:
        n = len(graph)
        color = [-1] * n   

        def bfs(start):
            q = deque([start])
            color[start] = 0  
            while q:
                node = q.popleft()
                for nei in graph[node]:
                    if color[nei] == color[node]: 
                        return False
                    if color[nei] == -1:        
                        color[nei] = 1 - color[node]
                        q.append(nei)
            return True

        for i in range(n):
            if color[i] == -1:
                if not bfs(i):
                    return False

        return True
